lib/stages/ReadGain.cpp

```diff
@@ -0,0 +1,240 @@
+        config.get<std::string>(unique_name, "updatable_config/gain_psr"),
+        std::bind(&ReadGain::update_gains_psr_callback, this, _1));
+}
+
+bool ReadGain::update_gains_frb_callback(nlohmann::json& json) {
+    {
+        std::lock_guard<std::mutex> lock(mux);
+        update_gains_frb = true;
+    }
+    cond_var.notify_all();
+
+    try {
+        _gain_dir_frb = json.at("frb_gain_dir");
+    } catch (std::exception& e) {
+        WARN("[FRB] Fail to read gain_dir {:s}", e.what());
+        return false;
+    }
+    INFO("[ReadGain] updated gain with {:s}============update_gains={:d}", _gain_dir_frb,
+         update_gains_frb);
+    return true;
+}
+
+bool ReadGain::update_gains_psr_callback(nlohmann::json& json) {
+    {
+        std::lock_guard<std::mutex> lock(mux);
+        update_gains_psr = true;
+    }
+    cond_var.notify_all();
+
+    try {
+        _gain_dir_psr = json.at("pulsar_gain_dir").get<std::vector<string>>();
+        INFO("[PSR] Updating gains from {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s} {:s}",
+             _gain_dir_psr[0].c_str(), _gain_dir_psr[1].c_str(), _gain_dir_psr[2].c_str(),
+             _gain_dir_psr[3].c_str(), _gain_dir_psr[4].c_str(), _gain_dir_psr[5].c_str(),
+             _gain_dir_psr[6].c_str(), _gain_dir_psr[7].c_str(), _gain_dir_psr[8].c_str(),
+             _gain_dir_psr[9].c_str());
+    } catch (std::exception const& e) {
+        WARN("[PSR] Fail to read gain_dir {:s}", e.what());
+        return false;
+    }
+    return true;
+}
+
+void ReadGain::main_thread() {
+
+    if (first_pass) {
+        first_pass = false;
+        uint8_t* frame =
+            wait_for_full_frame(metadata_buf, unique_name.c_str(), metadata_buffer_precondition_id);
+        if (frame == NULL)
+            goto end_loop;
+        stream_id_t stream_id = get_stream_id_t(metadata_buf, metadata_buffer_id);
+        freq_idx = bin_number_chime(&stream_id);
+        freq_MHz = freq_from_bin(freq_idx);
+        metadata_buffer_precondition_id =
+            (metadata_buffer_precondition_id + 1) % metadata_buf->num_frames;
+    }
+    mark_frame_empty(metadata_buf, unique_name.c_str(), metadata_buffer_id);
+    metadata_buffer_id = (metadata_buffer_id + 1) % metadata_buf->num_frames;
+    unregister_consumer(metadata_buf, unique_name.c_str());
+
+    while (!stop_thread) {
+        DEBUG("update_frb_gains={:d} update_psr_gains={:d} (0=false 1=true)", update_gains_frb,
+              update_gains_psr);
+
+        // FRB================================================
+        {
+            std::unique_lock<std::mutex> lock(mux);
+            while (!update_gains_frb) {
+                cond_var.wait(lock);
+            }
+        }
+        DEBUG("Going to update frb gain update_gains={:d}", update_gains_frb);
+
+        float* out_frame_frb =
+            (float*)wait_for_empty_frame(gain_frb_buf, unique_name.c_str(), gain_frb_buf_id);
+        if (out_frame_frb == NULL) {
+            goto end_loop;
```

Comment:
There is only one level of loop here, so `break;` will work. 
```suggestion
            break;
```